### DISCUSSION TITLE
pyproject.toml not bumping the version automatically

### DISCUSSION THREAD

**Discussion by UXabre** [renovatebot/renovate#28836] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

azure devops , and renovatebot as azure container job

### Please tell us more about your question or problem

We use requirements.txt for tracking dependencies, but we also want automatic version tracking. We found that we could use pep621 / pyproject.toml, but somehow, this does not bump the version.

I ran the bot in debug mode, and it sees 3 managers, pep621, pip_requirements and poetry. In the logs, it complains that it cannot parse poetry.

I'm not a python guy, just the devops guy...perhaps somebody has some pointers here?

### Logs (if relevant)

<details><summary>Logs</summary>

```

DEBUG: Matched 1 file(s) for manager pep621: pyproject.toml (repository=my-corp/python-project)
DEBUG: Matched 1 file(s) for manager pip_requirements: requirements.txt (repository=my-corp/python-project)
DEBUG: Matched 1 file(s) for manager poetry: pyproject.toml (repository=my-corp/python-project)
DEBUG: Poetry: error parsing pyproject.toml (repository=my-corp/python-project, packageFile=pyproject.toml)
       "err": {
         "message": "Schema error",
         "stack": "ZodError: Schema error\n    at Object.get error [as error] (/usr/local/renovate/node_modules/.pnpm/zod@3.23.4/node_modules/zod/lib/types.js:55:31)\n    at fromZodResult (/usr/local/renovate/lib/util/result.ts:54:67)\n    at Function.parse (/usr/local/renovate/lib/util/result.ts:552:12)\n    at Object.extractPackageFile (/usr/local/renovate/lib/modules/manager/poetry/extract.ts:19:36)\n    at extractPackageFile (/usr/local/renovate/lib/modules/manager/index.ts:75:9)\n    at getManagerPackageFiles (/usr/local/renovate/lib/workers/repository/extract/manager-files.ts:45:43)\n    at /usr/local/renovate/lib/workers/repository/extract/index.ts:57:28\n    at async Promise.all (index 2)\n    at extractAllDependencies (/usr/local/renovate/lib/workers/repository/extract/index.ts:54:26)\n    at extract (/usr/local/renovate/lib/workers/repository/process/extract-update.ts:140:28)\n    at extractDependencies (/usr/local/renovate/lib/workers/repository/process/index.ts:154:26)\n    at Object.renovateRepository (/usr/local/renovate/lib/workers/repository/index.ts:69:9)\n    at attributes.repository (/usr/local/renovate/lib/workers/global/index.ts:217:11)\n    at start (/usr/local/renovate/lib/workers/global/index.ts:202:7)\n    at /usr/local/renovate/lib/renovate.ts:18:22",
         "issues": {"tool": "Required"}
       }
DEBUG: manager extract durations (ms) (repository=my-corp/python-project)
       "managers": {"pip_requirements": 2, "poetry": 13, "pep621": 14}
DEBUG: Found pep621 package files (repository=my-corp/python-project)
DEBUG: Found pip_requirements package files (repository=my-corp/python-project)
DEBUG: Found 2 package file(s) (repository=my-corp/python-project)
 INFO: Dependency extraction complete (repository=my-corp/python-project, baseBranch=main)
       "stats": {
         "managers": {
           "pep621": {"fileCount": 1, "depCount": 1},
           "pip_requirements": {"fileCount": 1, "depCount": 6}
         },
         "total": {"fileCount": 2, "depCount": 7}
       }
DEBUG: hostRules: no authentication for pypi.org (repository=my-corp/python-project)


```

</details>

**Comment by RahulGautamSingh** [comment]
Seems like the structure of the `pyproject.toml` file you have doesn't match the structure we expect. 
To be more precise, the `Tool` field seems to be missing in the file.

We expect the following structure:
https://github.com/renovatebot/renovate/blob/ca23445c274fa6802f12ce564e8f7919a0d1a0f3/lib/modules/manager/poetry/schema.ts#L334-L336

You could compare it yourself and list the differences here, or link/paste the file here so we can inspect further.

**Reply by UXabre** [reply]
So, I double checked, it's not poetry. In renovate.json I've set:

```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "enabledManagers": ["pep621", "pip_requirements"]
}

```

This is the pyproject.toml
```
[build-system]
requires = ["hatchling"]
build-backend = "hatchling.build"

[project]
name = "<redacted>"
version = "1.0.2"
authors = [
  { name="<redacted>", email="<redacted>" },
]
description = "<redacted>"
readme = "README.md"
license = { file = "LICENSE"}
```

Next to this, I also have a requirements.txt. This seems to be updated with newer versions of dependencies, but it doesn't update the pyproject.toml. Or is this expected if these are two separate managers? Meaning... If I'd add the deps to pyproject.toml my life would be amazing?

**Reply by rarkins** [reply]
There are no versions to bump in the pyproject.toml you shared above. What are you hoping for Renovate to do?

**Reply by UXabre** [reply]
```
[project]
name = "<redacted>"
version = "1.0.2"
```

Is this not "bumpable"? Or is the field in the wrong place?
I expect this to be 1.0.3 (I've set bumpVersion to patch globally, so that is why it's not in the renovate.json)

**Reply by secustor** [reply]
This is the version of the current project. 
`bumpVersion` will ony bump this if there is an actual upgrade of dependencies. 
https://docs.renovatebot.com/configuration-options/#bumpversion

**Reply by UXabre** [reply]
Yeah, I hoped my dependencies, being tracked in requirements.txt would've done this. I'm no trying but putting the dependencies in the project file to see if it can tell what's going on

**Reply by rarkins** [reply]
So the dependencies are in requirements.txt but the project version is in pyproject.toml?

**Reply by UXabre** [reply, marked as answer]
Yes, that is the case... or better yet...was. As I now (probably) see the errors of my ways and have changed it track the deps in pyproject.toml and now it has started to bump the project version correctly as well as the deps versions

**Comment by secustor** [comment]
Poetry does not use the standard dependency definitions defined by Pep621, but rather use their one fields under tools 

https://python-poetry.org/docs/pyproject/#dependencies-and-dependency-groups

If you want to use the standard definitions you can for example use [PDM](https://pdm-project.org/latest/)

**Reply by rarkins** [reply]
Maybe we should be discarding it with a debug only

**Reply by secustor** [reply]
We should migrate poetry to a sub processor or pep621. There we have more context and then we have only one manager which looks at `pyproject.toml`

**Reply by rarkins** [reply]
I still don't like the idea of calling them all manager=pep621

**Comment by UXabre** [comment]
But, as far as i know, we don't use "poetry", it really is just a toml file with a [project] section, under which a name and version field are placed.

I'll check monday with the original developer if pdm is an option for him
